### PR TITLE
#2663 Remediation Workflow for Keptn 0.8

### LINKED ISSUE #2663: Remediation use case in Keptn 0.8
The goal of this issue is to support the remediation use case in Keptn 0.8

To reach this goal, the following changes need to be made:

## Remediation Service:

- The [isRemediationEnabled() function](https://github.com/keptn/keptn/blob/7bc1dafb6816c85fc56b03649f2fcd612c593835/remediation-service/handler/problem_open_handler.go#L97) currently checks, if the given stage has the `remediation_strategy` set to `automated` in the shipyard file. With the new [Shipyard format](https://github.com/keptn/spec/pull/36/files), we do not have this property anymore. Therefore, to check wether remediation is enabled, we can check if the `remediation.yaml` file is present in the config repo of the service.

- Instead of `sh.keptn.events.evaluation-done` events, the service needs to react to `sh.keptn.event.evaluation.finished` events (needs to be changed [here](https://github.com/keptn/keptn/blob/7bc1dafb6816c85fc56b03649f2fcd612c593835/remediation-service/handler/handler.go#L63)).

- Instead of sending `sh.keptn.event.start-evaluation` events, the service needs to send `sh.keptn.event.evaluation.triggered` events (needs to be changed [here](https://github.com/keptn/keptn/blob/7bc1dafb6816c85fc56b03649f2fcd612c593835/remediation-service/handler/handler.go#L335)).

- Previously, the service checked wether an evaluation.finished event is part of a remediation be checking wether the `teststrategy` was set to `real-user` (see [here](https://github.com/keptn/keptn/blob/7bc1dafb6816c85fc56b03649f2fcd612c593835/remediation-service/handler/evaluation_done_handler.go#L31)).  This property does not exist in the `evaluation.finished` events anymore. Instead of relying on this property, we can use the [getRemediationsByContext() function](https://github.com/keptn/keptn/blob/7bc1dafb6816c85fc56b03649f2fcd612c593835/remediation-service/handler/handler.go#L257) to check if an `evaluation.finished` event belongs to a remediation workflow using its `shkeptncontext` attribute.

## Integration Tests:

- Update the [test_self_healing.sh](https://github.com/keptn/keptn/blob/master/test/test_self_healing.sh) script to work with the current status on the master branch

- Update the [test_self_healing_scaling.sh](https://github.com/keptn/keptn/blob/master/test/test_self_healing_scaling.sh) script to work with the current status on the master branch (depends on https://github.com/keptn/keptn/issues/2558)

### PR DESCRIPTION
**Pull request by bacherfl**
Closes #2663 

As mentioned in Issue #2663, The integration test `test_self_healing_scaling.sh` depends on the completion of Issue #2558

### CODE REVIEW

File: remediation-service/handler/evaluation_done_handler.go [hunk 1]
```diff
@@ -11,95 +12,89 @@ import (
...
-		msg := "Remediation successful. Remediation actions resulted in evaluation result: " + evaluationDoneEventData.Result
+		msg := "Remediation successful. Remediation actions resulted in evaluation result: " + string(evaluationDoneEventData.Result)
 		eh.KeptnHandler.Logger.Info(msg)
 		return eh.Remediation.sendRemediationFinishedEvent(keptn.RemediationStatusSucceeded, keptn.RemediationResultPass, msg)
 	}
 
 	// get remediation.yaml
 	resource, err := eh.Remediation.getRemediationFile()
 	if err != nil {
+		_ = eh.Remediation.sendRemediationFinishedEvent(keptn.RemediationStatusErrored, keptn.RemediationResultFailed, err.Error())
```
Comment (johannes-b) [inline remediation-service/handler/evaluation_done_handler.go:52]: Can we also log the error here, as done in line 59: `eh.KeptnHandler.Logger.Error(err.Error())` - to tell the user that the `remediation.yaml` was not found.

File: remediation-service/handler/problem_open_handler.go [hunk 2]
```diff
@@ -32,26 +31,32 @@ func (eh *ProblemOpenEventHandler) HandleEvent() error {
...
+		msg := fmt.Sprintf("Remediation disabled for service %s in project %s in stage %s", problemEvent.Service, problemEvent.Project, problemEvent.Stage)
+		eh.KeptnHandler.Logger.Info(msg)
+		_ = eh.Remediation.sendRemediationFinishedEvent(keptn.RemediationStatusErrored, keptn.RemediationResultFailed, msg)
 		return nil
 	}
 
 	// get remediation.yaml
 	resource, err := eh.Remediation.getRemediationFile()
 	if err != nil {
+		_ = eh.Remediation.sendRemediationFinishedEvent(keptn.RemediationStatusErrored, keptn.RemediationResultFailed, err.Error())
```
Comment (johannes-b) [inline remediation-service/handler/problem_open_handler.go:51]: Like above, should we log an error here?

**Comment by johannes-b** [comment]
From my perspective, the PR looks good (except for the two minor enhancements regarding logging)

File: remediation-service/handler/evaluation_done_handler.go (hunk 1 above)
Comment (bacherfl) [reply remediation-service/handler/evaluation_done_handler.go:53]: good catch - has been updated

File: remediation-service/handler/problem_open_handler.go (hunk 2 above)
Comment (bacherfl) [reply remediation-service/handler/problem_open_handler.go:52]: good catch - has been updated

**Review by johannes-b — APPROVED** [review]
lgtm

**Pull request merged**